src/Abp/EntityHistory/EntityHistoryInterceptor.cs

```diff
@@ -0,0 +1,28 @@
+﻿using Castle.DynamicProxy;
+using System.Linq;
+using System.Reflection;
+
+namespace Abp.EntityHistory
+{
+    internal class EntityHistoryInterceptor : IInterceptor
+    {
+        public IEntityChangeSetReasonProvider ReasonProvider { get; set; }
+
+        public EntityHistoryInterceptor()
+        {
+            ReasonProvider = NullEntityChangeSetReasonProvider.Instance;
+        }
+
+        public void Intercept(IInvocation invocation)
+        {
+            var methodInfo = invocation.MethodInvocationTarget;
+            var useCaseAttribute = methodInfo.GetCustomAttributes<UseCaseAttribute>(true).FirstOrDefault()
+                  ?? methodInfo.DeclaringType.GetCustomAttributes<UseCaseAttribute>(true).First();
```

Comment:
What if ``useCaseAttribute`` is not defined for the method or declaring type?
This is possible if ``entityHistoryConfiguration.Selectors.Any(selector => selector.Predicate(type))`` is true in ``EntityHistoryInterceptorRegistrar.ShouldIntercept`` method.

Reply:
We *could* use the method name.

However, I will remove that condition in `ShouldIntercept` as `Selectors` are for a different purpose: selecting entities as an alternative (or an equivalent) to applying the `[Audited]` attribute on entities. That may get confusing if `Selectors` were also an alternative to applying the `[UseCase]` attribute, especially since applying the `[Audited]` attribute on a service class already has a defined behavior.

Reply:
There is still a problem if a class has two methods one with UseCase other without UseCase (and declaring type has not UseCase). When you call the method without use case, it throws the following exception:

System.InvalidOperationException : Sequence contains no elements.

Stacktrace:

![image](https://user-images.githubusercontent.com/1210527/34660616-c9265bfa-f454-11e7-9d87-aec9ad04918a.png)

I fixed it by this commit (and included unit test): https://github.com/aspnetboilerplate/aspnetboilerplate/commit/413ab18769cccd705c2413f397cebe42bba3e417
(also used )

Reply:
Right, thanks for handling that!